onnxruntime/core/providers/cuda/cuda_allocator.cc

```diff
@@ -76,7 +76,21 @@ void* CUDAExternalAllocator::Alloc(size_t size) {
 }
 
 void CUDAExternalAllocator::Free(void* p) {
+  std::lock_guard<OrtMutex> lock(lock_);
   free_(p);
+  auto it = reserved_.find(p);
+  if (it != reserved_.end()) {
+    reserved_.erase(it);
+    if (empty_cache_) empty_cache_();
+  }
+}
+
+void* CUDAExternalAllocator::Reserve(size_t size) {
+  std::lock_guard<OrtMutex> lock(lock_);
+  void* p = Alloc(size);
```

Comment:
Alloc(size) doesn't need to be guarded under lock ?

Reply:
The lock here and in Free below is mainly for the std::unordered_map. I think std::unordered_map is not thread safe.